Multiversion client tests

soltysh opened the pull request:
@caesarxuchao this is a follow up to the discussion we've had while I was working on the `JobTemplate`, esp. the part about introducing `batch/v2alpha1` (#21675). I've split this from #25475 not to block progress on `ScheduledJobs`, so look only at the last commit. We need some sort of solution for the multi-version client testing and I'm willing to work on it but will need your help. 

In the aforementioned commit I've extended `simpletest_client.go`'s `Config with`GroupVersion`which allows you to specify which version you want to test. This required some additional changes in`testapi.go`to be able to specify which specific version we're testing. See`jobs_test.go`, since with those changes I was easily able to test all possible versions`Job` exists in.

I'm open to suggestions how to improve the code, this is just a dirty plumbing I did to make the test work, while mostly focusing on ScheduledJobs.

@erictune also mentioned there's a plan to get rid of manually writing those clients, in that case it might make sense just to ignore the problems for now and wait for the generated clients make their way. The only question is when will that happen?

[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/.github/PULL_REQUEST_TEMPLATE.md?pixel)]()

soltysh committed:
ScheduledJob client

soltysh committed:
Adding github.com/robfig/cron for cron parsing

soltysh committed:
ScheduledJob validation & defaults

soltysh committed:
ScheduledJob api leftovers

soltysh committed:
ScheduledJob api leftovers - generated

soltysh committed:
Enable multi-version tests in client tests

deads2k commented on pkg/api/testapi/testapi.go:
```diff
@@ -216,8 +225,11 @@ func (g TestGroup) InternalTypes() map[string]reflect.Type {
-func (g TestGroup) Codec() runtime.Codec {
+func (g TestGroup) Codec(gv *unversioned.GroupVersion) runtime.Codec {
 	if serializer.Serializer == nil {
+		if g.IsEnabledVersion(gv) {
```
If I requested a specific version and I can't have it, shouldn't I fail?

soltysh replied:
Yeah the overall lack of error handling or panic (it's just tests) is something I've struggled while working on this.

deads2k commented on pkg/api/testapi/testapi.go:
```diff
@@ -191,7 +191,16 @@ func init() {
+
+func (g TestGroup) IsEnabledVersion(v *unversioned.GroupVersion) bool {
+	for _, gv := range g.externalGroupVersions {
+		if gv.String() == v.String() {
```
shouldn't need to convert to string, just `*v` right?

deads2k commented on pkg/api/testapi/testapi.go:
```diff
@@ -191,7 +191,16 @@ func init() {
+	return "application/json", g.GroupVersion(), g.Codec(g.GroupVersion())
+}
+
+func (g TestGroup) IsEnabledVersion(v *unversioned.GroupVersion) bool {
```
why is this a pointer?  Is it supposed to be handling `nil` somehow?

deads2k commented on pkg/api/testapi/testapi.go:
```diff
@@ -293,14 +305,18 @@ func (g TestGroup) SelfLink(resource, name string) string {
 		// TODO: switch back once we have proper multiple group support
 		// path = "/apis/" + g.Group + "/" + Version(group...)
-		path = "/apis/" + g.externalGroupVersions[0].Group + "/" + g.externalGroupVersions[0].Version
+		path = "/apis/" + groupVersion.Group + "/" + groupVersion.Version
```
`gvr.Group`, `gvr.Version`?

deads2k replied:
> gvr.Group, gvr.Version?

nm

deads2k commented on pkg/client/unversioned/testclient/simple/simple_testclient.go:
```diff
@@ -71,6 +71,9 @@ type Client struct {
 	ResourceGroup string
+
+	// Specific GroupVersion that should be used for testing.
+	GroupVersion *unversioned.GroupVersion
```
What does nil mean here?

soltysh replied:
Use the default available.

deads2k commented on pkg/client/unversioned/testclient/simple/simple_testclient.go:
```diff
@@ -108,6 +111,14 @@ func (c *Client) Setup(t *testing.T) *Client {
 }
 
+func (c *Client) groupVersion(group testapi.TestGroup) *unversioned.GroupVersion {
+	if c.GroupVersion != nil && group.IsEnabledVersion(c.GroupVersion) {
```
does this mean that when I request a specific version, if its not enabled I'll get back a client that points to a different group version?

soltysh replied:
Unfortunately, other option is to panic/error like in the other case.

caesarxuchao commented:
@soltysh could you remind me the context? I don't remember why can't we use the KUBE_TEST_API_VERSIONS environment variable to control what version to test.

soltysh commented:
@caesarxuchao the problem with enabling just `batch/v2alpha1` with `KUBE_TEST_API_VERSIONS` is that when that happens `batch/install.go` still installs all the versions at once. Since `batch/v1` is the preferred version it is installed along with `batch/v2alpha1`. In that cases we need to force specific version in tests, especially with `ScheduledJobs` since these are available only in `batch/v2alpha1`. In that case the tests must be conditional, iow. just for that specific version, not just group. Generally this is the first place where we actually introduce 2nd version in any of the groups.

caesarxuchao commented:
> the problem with enabling just batch/v2alpha1 with KUBE_TEST_API_VERSIONS is that when that happens batch/install.go still installs all the versions at once. Since batch/v1 is the preferred version it is installed along with batch/v2alpha1

I must have missed something: it seems if you only put `batch/v2alpha1` in `KUBE_TEST_API`, you will only get that version from the testapi. (https://github.com/kubernetes/kubernetes/blob/master/pkg/api/testapi/testapi.go#L94)
In the tests that should only be run for v2alpha1, you can add the following `if` statement at the beginning of your test:
```
if testapi.Batch.GroupVersion != "batch/v2alpha1" {
  return
}
```

Sorry I cannot recall our discussion in the other PR..

soltysh commented:
> I must have missed something: it seems if you only put batch/v2alpha1 in KUBE_TEST_API, you will only get that version from the testapi. (https://github.com/kubernetes/kubernetes/blob/master/pkg/api/testapi/testapi.go#L94)

That's not quite the way it works, iirc. I can't remember the full details, since that was a while, but in general this line will allow only this version, but since [batch/install.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/testapi/testapi.go#L43) is called it installs all the versions internally and that is later the root cause of all the problems. 

I think the case was for the situation where you don't specify any of batch versions and then all are installed, which is why the tests were failing (and one of the groups installs extensions but no batch, see the first group in [KUBE_TEST_API_VERSIONS](https://github.com/kubernetes/kubernetes/blob/master/hack/test-go.sh#L61).

caesarxuchao commented:
Just to reiterate, so you agree that if only batck/v1beta2 is specified in "KUBE_API_TEST", then there will be no problem?

> and one of the groups installs extensions but no batch, see the first group in KUBE_TEST_API_VERSIONS

In that case, calling `testapi.Batch.GroupVersion` should give you "batch/v1beta1", and the `if` statement I mentioned should help you skip the tests.

soltysh commented:
Hmmm... lemme check that again. I think there was more than just that.

soltysh commented:
@caesarxuchao so here's the problem I've stumbled upon, when I've enabled just `batch/v1` I'm getting following test error:
```
--- FAIL: TestDiscoveryAtAPIS (0.42s)
	Error Trace:	master_test.go:475
	Error:		Not equal: []unversioned.GroupVersionForDiscovery{unversioned.GroupVersionForDiscovery{GroupVersion:"batch/v1", Version:"v1"}} (expected)
			        != []unversioned.GroupVersionForDiscovery{unversioned.GroupVersionForDiscovery{GroupVersion:"batch/v1", Version:"v1"}, unversioned.GroupVersionForDiscovery{GroupVersion:"batch/v2alpha1", Version:"v2alpha1"}} (actual)
			
			Diff:
			--- Expected
			+++ Actual
			@@ -1,3 +1,4 @@
			-([]unversioned.GroupVersionForDiscovery) (len=1 cap=1) {
			- (unversioned.GroupVersionForDiscovery) groupVersion:"batch/v1" version:"v1" 
			+([]unversioned.GroupVersionForDiscovery) (len=2 cap=4) {
			+ (unversioned.GroupVersionForDiscovery) groupVersion:"batch/v1" version:"v1" ,
			+ (unversioned.GroupVersionForDiscovery) groupVersion:"batch/v2alpha1" version:"v2alpha1" 
			 }
		
	Error Trace:	master_test.go:506
	Error:		Not equal: []unversioned.GroupVersionForDiscovery{unversioned.GroupVersionForDiscovery{GroupVersion:"batch/v1", Version:"v1"}} (expected)
			        != []unversioned.GroupVersionForDiscovery{unversioned.GroupVersionForDiscovery{GroupVersion:"batch/v1", Version:"v1"}, unversioned.GroupVersionForDiscovery{GroupVersion:"batch/v2alpha1", Version:"v2alpha1"}} (actual)
			
			Diff:
			--- Expected
			+++ Actual
			@@ -1,3 +1,4 @@
			-([]unversioned.GroupVersionForDiscovery) (len=1 cap=1) {
			- (unversioned.GroupVersionForDiscovery) groupVersion:"batch/v1" version:"v1" 
			+([]unversioned.GroupVersionForDiscovery) (len=2 cap=4) {
			+ (unversioned.GroupVersionForDiscovery) groupVersion:"batch/v1" version:"v1" ,
			+ (unversioned.GroupVersionForDiscovery) groupVersion:"batch/v2alpha1" version:"v2alpha1" 
			 }
```
I remember I was trying to fix it, but failed badly. See that although `batch/v1` is forced through `KUBE_TEST_API_VERSIONS` both versions get actually installed. Without the multiple versions (I've done so far) only the latest registered version got in. In all cases (for every batch version) that used to be `batch/v2alpha1`, being the last one. I'm open to any ideas.

soltysh commented:
Identical problems in `pkg/master/` are when just running `batch/v2alpha1`:
```
--- FAIL: TestDiscoveryAtAPIS (0.49s)
	Error Trace:	master_test.go:475
	Error:		Not equal: []unversioned.GroupVersionForDiscovery{unversioned.GroupVersionForDiscovery{GroupVersion:"batch/v2alpha1", Version:"v2alpha1"}} (expected)
			        != []unversioned.GroupVersionForDiscovery{unversioned.GroupVersionForDiscovery{GroupVersion:"batch/v1", Version:"v1"}, unversioned.GroupVersionForDiscovery{GroupVersion:"batch/v2alpha1", Version:"v2alpha1"}} (actual)
			
			Diff:
			--- Expected
			+++ Actual
			@@ -1,2 +1,3 @@
			-([]unversioned.GroupVersionForDiscovery) (len=1 cap=1) {
			+([]unversioned.GroupVersionForDiscovery) (len=2 cap=4) {
			+ (unversioned.GroupVersionForDiscovery) groupVersion:"batch/v1" version:"v1" ,
			  (unversioned.GroupVersionForDiscovery) groupVersion:"batch/v2alpha1" version:"v2alpha1" 
		
	Error Trace:	master_test.go:506
	Error:		Not equal: []unversioned.GroupVersionForDiscovery{unversioned.GroupVersionForDiscovery{GroupVersion:"batch/v2alpha1", Version:"v2alpha1"}} (expected)
			        != []unversioned.GroupVersionForDiscovery{unversioned.GroupVersionForDiscovery{GroupVersion:"batch/v1", Version:"v1"}, unversioned.GroupVersionForDiscovery{GroupVersion:"batch/v2alpha1", Version:"v2alpha1"}} (actual)
			
			Diff:
			--- Expected
			+++ Actual
			@@ -1,2 +1,3 @@
			-([]unversioned.GroupVersionForDiscovery) (len=1 cap=1) {
			+([]unversioned.GroupVersionForDiscovery) (len=2 cap=4) {
			+ (unversioned.GroupVersionForDiscovery) groupVersion:"batch/v1" version:"v1" ,
			  (unversioned.GroupVersionForDiscovery) groupVersion:"batch/v2alpha1" version:"v2alpha1" 
```

soltysh commented:
Maybe I'm missing a better solution here, I don't deny it, that's why I've opened this issue to discuss possible approaches :)

soltysh commented:
After second thought, we could install in the testapi only the version that is specified and ignore everything else. I need to give it a try, since it looks like it might help here.

soltysh commented:
Current state is that I've reverted the changes I've did in `pkg/api/testapi/testapi.go` to a single `externalGroupVersion`, like before and modified `pkg/master/master_test.go` to pass the tests. I've introduced 3rd `KUBE_TEST_API_VERSIONS` group with `batch/v2alpha`. The only problem I'm trying to address right now is the serialization tests. Hopefully I'll be able to fix it and re-do this PR with a better approach (or probably open a new one).

caesarxuchao commented:
Thanks @soltysh! Yes, those master_test should only be run for a certain KUBE_API_TEST_VERSION.
